libbeat/reader/multiline/message_buffer.go

```diff
@@ -120,11 +129,18 @@ func (b *messageBuffer) addLine(m reader.Message) {
 // finalize writes the existing content into the returned message and resets all reader variables.
 func (b *messageBuffer) finalize() reader.Message {
 	if b.truncated > 0 {
-		b.message.AddFlagsWithKey("log.flags", "truncated")
+		truncatedLogRate.Do(func() {
```

Comment:
I like the ergonomics of the rate limited logger, but thinking into a future in which this is widely used on the hot event path, do we know what the performance cost of each instance of the rate limited logger is?

Stepping back a bit, the message is mostly just telling us that something happened, and to me that is a hint that it should just be a metric counter instead, and not a rate limited log line.

Reply:
> I like the ergonomics of the rate limited logger, but thinking into a future in which this is widely used on the hot event path, do we know what the performance cost of each instance of the rate limited logger is?

I'm somewhat skeptical about the current logger code, so I'm still experimenting with it. The performance cost includes a pointer per logger instance and a mutex lock per message. In a multi-threaded scenario, this would serialize execution. However, I would like to add benchmarks for this scenario.

> Stepping back a bit, the message is mostly just telling us that something happened, and to me that is a hint that it should just be a metric counter instead, and not a rate limited log line.

I'm happy with this outcome. Do you have any directions on how to go about adding the metric here? Thanks.

Reply:
This would probably best fit as an input level metric. This would probably require you to return an error detectable at the input level to allow the counter to increment.

Filestream input metrics were added in https://github.com/elastic/beats/pull/35529 as a reference.

There are libbeat metrics from before this existed that are specific to filebeat as a less preferred alternative.

https://github.com/elastic/beats/blob/4942592a8a272b5355c0be531e5f052b2c7a1d56/filebeat/input/log/harvester.go#L69

https://github.com/elastic/beats/blob/4942592a8a272b5355c0be531e5f052b2c7a1d56/libbeat/monitoring/report/log/log.go#L52

Reply:
Thanks Craig, and sorry for the delay in addressing this. I changed the log into a metric as you suggested. Since we set a `log.flags` directly on the message fields I decided to just check for that instead of a sentinel error.